### ISSUE TITLE
Automate local S3 ETL in cloud

### ISSUE THREAD

**Issue by e-belfer** [catalyst-cooperative/pudl-usage-metrics#147]
### Overview

What is the problem we're solving? For very simple items, this can be encapsulated in the success criteria.

Once #140 is merged, we'll have a local functioning ETL that processes weekly slices of S3 logs. The next step is to get this working in the cloud and on Github actions. For now, we'll stick to using Postgres/CloudSQL, though we'll eventually want to move to BigQuery.

### Success Criteria

How will we know that we're done?
We have a scheduled Github action that runs the latest partition of S3 metrics once a week and updates the CloudSQL database in GCS and in Superset.

```[tasklist]
### Next steps
- [x] Create alembic schema for DB
- [x] Update run_data_update to run last week's partition
- [x] If alembic schema changes - update postgres db
- [x] Get Postgres io manager working for s3 runs
- [x] Get `run_data_update.py` working as expected
- [x] Handle inevitable GCS configuration problems
- [x] Hook up to superset
- [x] Add tempdir / non-temp dir distinction to docs
- [x] Add `partition_key` to `load_input` for SQLiteIOManager and PostgresIOManager
- [x] Run backfill for all S3 data in production
- [x] Handle missing partitions
- [x] Add bogon boolean
- [x] Test GHA to run weekly partition
```

**Comment by bendnorman** [comment]
We can do datasource-level permissions on superset which means we can use our deployment at superset.catalyst.coop for pudl usage metrics. We should have 3 roles:
- A couple of catalyst people should have the Admin role
- We should create a role for all Catalyst folks. Probably just use the permissions in the predefined Alpha role
- Our `GammaSQLLab` role should be renamed to `PUDLGammaSQLLab` make sure they can only access the PUDL db.

It's probably easiest to write the metrics logs to a duckdb file so we can upload it to `gs://superset.catalyst.coop`.

**Comment by jdangerx** [comment]
> It's probably easiest to write the metrics logs to a duckdb file so we can upload it to `gs://superset.catalyst.coop`.

Ah, instead of shoving things into Postgres/Cloud SQL? That does seem easier, I'll do that.

**Comment by bendnorman** [comment]
@e-belfer and I chatted yesterday about which storage backend to use for our usage metrics. We decided using BigQuery in production and DuckDB for local development would be ideal. I don't think using duckdb in production would work because we can't do remote writes to a duckdb file so we'd have to download the entire duckdb file, write new data and upload the whole thing again. With BigQuery we can just write new data to the database. Duckdb would be nice because it's faster and has stronger types than SQLite but it's not required.

While BQ and Duckdb are our ideal setup, we think our existing setup of SQLite for local development and Cloud SQL for production is sufficient. Here are the tasks that need to be completed to get the s3 logs into production:

- Connect our usage metrics Cloud SQL instance to our  Superset Cloud Run deployment. This should be pretty straightforward. I did this to connect superset to it's backend database.
- Set up alembic to manage schema changes. Right now the `SQLiteIOManager` and `PostgresManager` have a clobber argument that can wipe out the database and recreate the database tables. This is a pretty brute-force method of managing schema changes so we should use alembic instead!
- Right now we [white list](https://github.com/catalyst-cooperative/pudl-usage-metrics/blob/0804f361fd2a9e000078809abdb37ef16c101058/.github/workflows/load-metrics.yml#L56-L58) the github runner IP in `load-metrics` to connect to the Cloud SQL database. This is ok but the proper way to connect to a Cloud SQL db is using the Cloud SQL proxy. It [looks like some people](https://github.com/GoogleCloudPlatform/cloud-sql-proxy/issues/1989) have managed to use cloud SQL proxy in a github action. I wonder if down the line it makes more sense for the github action to just build a docker image and execute it in Google Batch. What do you think @jdangerx?

Some things that are out of scope for this round:
- Set up duckdb for local development
- Set up bigquery for production development
- Move weekly processing into a Google product instead of a GitHub action.

**Comment by e-belfer** [comment]
@bendnorman and @jdangerx would love your thoughts on this!

Right now I have dynamic weekly partitions for each asset (raw, core, out). All is well between raw and core tables, but because the `load_input` setting for both our SQL IO managers doesn't recognize partition keys, it's reading in the entire `core_s3_logs` table to process for each `out_s3_logs` partition, which is creating duplicated data (each `out_s3_log` partition contains all processed data) and causing errors. I can solve this in a few ways:

- somehow get `load_input` to look for a `DateTime` key and filter values to match the partition key (only read in the desired slice of `core_s3_logs`). Timezones will make this complicated, since the start date of our records is often before the day, but we could probably figure out how to resolve this. There isn't a consistently named time field across S3 and non-S3 sources right now, but we could also resolve this.
- only keep `core_s3_logs`, and do the very minor filtering I currently do for `out_s3_logs` in Superset.
- something else I'm not seeing

**Comment by bendnorman** [comment]
Ah great point. The dagster docs don't give an example of a database IO manager that can handle partitions but they give some [suggestions](https://docs.dagster.io/concepts/io-management/io-managers#handling-partitioned-assets):

> I/O managers can be written to handle [partitioned](https://docs.dagster.io/concepts/partitions-schedules-sensors/partitions) assets. For a partitioned asset, each invocation of handle_output will (over)write a single partition, and each invocation of load_input will load one or more partitions. When the I/O manager is backed by a filesystem or object store, then each partition will typically correspond to a file or object. When it's backed by a database, then each partition will typically correspond to a range of rows in a table that fall within a particular window.

> To handle partitions in an custom I/O manager, you'll need to determine which partition you're dealing with when you're storing an output or loading an input. For this, [OutputContext](https://docs.dagster.io/_apidocs/io-managers#dagster.OutputContext) and [InputContext](https://docs.dagster.io/_apidocs/io-managers#dagster.InputContext) have a asset_partition_key property:

We could add an `asset_partition_key` column to all the tables in the database. I think this avoids the time zone and consistent date column name issues you mentioned. The IO manager would looks something like this

```python
def handle_output(context, obj: pd.DataFrame):
    if context.has_partition_key:
        obj["partition_key"] = context.asset_partition_key
    # write the df to the db

def load_input(context):
    asset_name = ... # get the desired asset name from context
    if context.has_partition_key:
         partition_key = context.asset_partition_key
         query = f"SELECT * FROM {asset_name} WHERE partition_key = '{partition_key}'"
         obj = pd.read_sql(query, conn)
    else:
          # handle non partitioned asset
    return obj
```

**Comment by e-belfer** [comment]
> We could add an asset_partition_key column to all the tables in the database.

This is a great solution, I'll go ahead and implement it!
